File: linuxdo-guest-browser/pycharm/src/main/java/studio/lexiao/linuxdo/ReaderHistory.java
package studio.lexiao.linuxdo;

import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class ReaderHistory {
    static final int MAX_ENTRIES = 60;
    private static final int MAX_TITLE_LENGTH = 200;
    private static final int MAX_URL_LENGTH = 2048;
    private static final Pattern TOPIC_PATH = Pattern.compile("^/t/(?:([^/]+)/)?([1-9][0-9]*)(?:/.*)?$");

    record Entry(String url, String title, long visitedAt) {
    }

    private ReaderHistory() {
    }

    static Entry create(String rawUrl, String rawTitle, long visitedAt) {
        String url = normalizeUrl(rawUrl);
        if (url == null || visitedAt <= 0) return null;
        String title = String.valueOf(rawTitle == null ? "" : rawTitle)
                .replaceAll("[\\r\\n\\t]+", " ")
                .replaceAll("\\s+", " ")
                .replaceFirst("^\\([0-9]+\\)\\s*", "")
                .replaceAll("\\s+-\\s+(?:LINUX DO|搞七捻三)\\s*$", "")
                .trim();
        if (url.equals("https://linux.do/latest")) title = "最新主题";
        else if (url.equals("https://linux.do/top")) title = "本周热门";
        else if (url.equals("https://linux.do/categories")) title = "浏览分类";
        else if (url.startsWith("https://linux.do/search")) title = "搜索结果";
        if (title.isEmpty()) title = "公开页面";
        if (title.length() > MAX_TITLE_LENGTH) title = title.substring(0, MAX_TITLE_LENGTH);
        return new Entry(url, title, visitedAt);
    }

    static List<Entry> add(List<Entry> current, Entry candidate) {
        ArrayList<Entry> result = new ArrayList<>();
        if (candidate != null) result.add(candidate);
        if (current != null) {
            for (Entry entry : current) {
                if (entry == null || candidate != null && candidate.url().equals(entry.url())) continue;
                Entry normalized = create(entry.url(), entry.title(), entry.visitedAt());
                if (normalized != null) result.add(normalized);
                if (result.size() >= MAX_ENTRIES) break;
            }
        }
        return result;
    }

    static List<Entry> updateTitle(List<Entry> current, String rawUrl, String rawTitle) {
        String url = normalizeUrl(rawUrl);
        if (url == null || current == null || current.isEmpty()) return current == null ? List.of() : new ArrayList<>(current);
        ArrayList<Entry> result = new ArrayList<>();
        for (Entry entry : current) {
            if (entry == null) continue;
            Entry updated = url.equals(entry.url()) ? create(entry.url(), rawTitle, entry.visitedAt()) : entry;
            Entry normalized = updated == null ? null : create(updated.url(), updated.title(), updated.visitedAt());
            if (normalized != null) result.add(normalized);
            if (result.size() >= MAX_ENTRIES) break;
        }
        return result;
    }

    static String serialize(List<Entry> entries) {
        StringBuilder result = new StringBuilder();
        if (entries == null) return "";
        int count = 0;
        for (Entry entry : entries) {
            Entry normalized = entry == null ? null : create(entry.url(), entry.title(), entry.visitedAt());
            if (normalized == null) continue;
            if (count > 0) result.append('\n');
            result.append(normalized.visitedAt()).append('\t')
                    .append(encode(normalized.title())).append('\t')
                    .append(encode(normalized.url()));
            count += 1;
            if (count >= MAX_ENTRIES) break;
        }
        return result.toString();
    }

    static List<Entry> parse(String value) {
        ArrayList<Entry> entries = new ArrayList<>();
        if (value == null || value.isBlank()) return entries;
        for (String line : value.split("\\n")) {
            String[] fields = line.split("\\t", -1);
            if (fields.length != 3) continue;
            try {
                long visitedAt = Long.parseLong(fields[0]);
                Entry entry = create(decode(fields[2]), decode(fields[1]), visitedAt);
                if (entry != null && entries.stream().noneMatch(current -> current.url().equals(entry.url()))) {
                    entries.add(entry);
                }
            } catch (IllegalArgumentException ignored) {
                // Ignore a damaged local history entry without losing valid entries.
            }
            if (entries.size() >= MAX_ENTRIES) break;
        }
        return entries;
    }

    private static String normalizeUrl(String rawUrl) {
        if (rawUrl == null || rawUrl.length() > MAX_URL_LENGTH) return null;
        try {
            URI uri = URI.create(rawUrl.trim());
            if (!"https".equalsIgnoreCase(uri.getScheme()) || !"linux.do".equalsIgnoreCase(uri.getHost())
                    || uri.getUserInfo() != null) return null;
            String path = uri.getPath() == null || uri.getPath().isBlank() ? "/latest" : uri.getPath();
            String lowerPath = path.toLowerCase();
            if (lowerPath.startsWith("/login") || lowerPath.startsWith("/signup")
                    || lowerPath.startsWith("/session") || lowerPath.startsWith("/auth")
                    || lowerPath.startsWith("/cdn-cgi/")) return null;
            String query = uri.getQuery();
            if (query != null && query.toLowerCase().contains("__cf_chl")) return null;
            Matcher topic = TOPIC_PATH.matcher(path);
            if (topic.matches()) {
                String slug = topic.group(1) == null ? "topic" : topic.group(1);
                path = "/t/" + slug + "/" + topic.group(2);
                query = null;
            } else if (lowerPath.matches("/(?:latest|top|categories)/?")) {
                path = lowerPath.replaceAll("/$", "");
                query = null;
            }
            return new URI("https", null, "linux.do", -1, path, query, null).toString();
        } catch (IllegalArgumentException | URISyntaxException ignored) {
            return null;
        }
    }

    private static String encode(String value) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(value.getBytes(StandardCharsets.UTF_8));
    }

    private static String decode(String value) {
        byte[] bytes = Base64.getUrlDecoder().decode(value);
        if (bytes.length > MAX_URL_LENGTH * 4) throw new IllegalArgumentException("history field too large");
        return new String(bytes, StandardCharsets.UTF_8);
    }
}

File: linuxdo-guest-browser/vscode/src/reader-history.js
'use strict';

const MAX_ENTRIES = 60;
const MAX_TITLE_LENGTH = 200;
const MAX_URL_LENGTH = 2048;
const SITE_ORIGIN = 'https://linux.do';

function createHistoryEntry(action, title, explicitUrl, visitedAt = Date.now()) {
  const normalizedAction = normalizeAction(action);
  const url = normalizePublicUrl(explicitUrl || actionUrl(normalizedAction));
  const timestamp = Number(visitedAt);
  if (!normalizedAction || !url || !Number.isInteger(timestamp) || timestamp <= 0) return undefined;
  return { url, title: normalizeTitle(title || actionTitle(normalizedAction)), visitedAt: timestamp, action: normalizedAction };
}

function addHistoryEntry(current, candidate) {
  const result = [];
  if (candidate) result.push(candidate);
  for (const rawEntry of Array.isArray(current) ? current : []) {
    const entry = normalizeStoredEntry(rawEntry);
    if (!entry || candidate?.url === entry.url) continue;
    result.push(entry);
    if (result.length >= MAX_ENTRIES) break;
  }
  return result;
}

function normalizeStoredHistory(value) {
  const result = [];
  for (const rawEntry of Array.isArray(value) ? value : []) {
    const entry = normalizeStoredEntry(rawEntry);
    if (!entry || result.some((current) => current.url === entry.url)) continue;
    result.push(entry);
    if (result.length >= MAX_ENTRIES) break;
  }
  return result;
}

function normalizeStoredEntry(value) {
  if (!value || typeof value !== 'object') return undefined;
  return createHistoryEntry(value.action, value.title, value.url, Number(value.visitedAt));
}

function normalizeAction(value) {
  if (!value || typeof value !== 'object') return undefined;
  if (value.type === 'topic') {
    const id = Number(value.id);
    const slug = String(value.slug || '').trim();
    if (!Number.isInteger(id) || id <= 0 || !/^[A-Za-z0-9_-]{1,200}$/.test(slug)) return undefined;
    return { type: 'topic', id, slug };
  }
  if (value.type === 'category') {
    const id = Number(value.id);
    if (!Number.isInteger(id) || id <= 0) return undefined;
    return { type: 'category', id, name: normalizeTitle(value.name || '分类') };
  }
  if (value.type === 'search') {
    const query = String(value.query || '').replace(/[\r\n\t]+/g, ' ').trim().slice(0, 100);
    return query ? { type: 'search', query } : undefined;
  }
  const view = String(value.view || '');
  return value.type === 'view' && ['latest', 'top', 'categories'].includes(view)
    ? { type: 'view', view }
    : undefined;
}

function actionUrl(action) {
  if (!action) return '';
  if (action.type === 'topic') return `${SITE_ORIGIN}/t/${action.slug}/${action.id}`;
  if (action.type === 'category') return `${SITE_ORIGIN}/c/${action.id}`;
  if (action.type === 'search') return `${SITE_ORIGIN}/search?q=${encodeURIComponent(action.query)}`;
  if (action.view === 'top') return `${SITE_ORIGIN}/top`;
  if (action.view === 'categories') return `${SITE_ORIGIN}/categories`;
  return `${SITE_ORIGIN}/latest`;
}

function actionTitle(action) {
  if (!action) return '公开页面';
  if (action.type === 'category') return `${action.name} · 分类主题`;
  if (action.type === 'search') return `搜索：${action.query}`;
  if (action.type === 'topic') return 'LINUX DO 公开主题';
  if (action.view === 'top') return '本周热门';
  if (action.view === 'categories') return '浏览分类';
  return '最新主题';
}

function normalizeTitle(value) {
  const title = String(value || '')
    .replace(/[\r\n\t]+/g, ' ')
    .replace(/\s+/g, ' ')
    .replace(/^\([0-9]+\)\s*/, '')
    .replace(/\s+-\s+(?:LINUX DO|搞七捻三)\s*$/i, '')
    .trim();
  return (title || '公开页面').slice(0, MAX_TITLE_LENGTH);
}

function normalizePublicUrl(value) {
  const input = String(value || '').trim();
  if (!input || input.length > MAX_URL_LENGTH) return undefined;
  try {
    const url = new URL(input, SITE_ORIGIN);
    if (url.origin !== SITE_ORIGIN || url.username || url.password) return undefined;
    if (/^\/(?:login|signup|session|auth)(?:\/|$)/i.test(url.pathname)) return undefined;
    if (/^\/cdn-cgi\//i.test(url.pathname) || [...url.searchParams.keys()].some((key) => key.toLowerCase().startsWith('__cf_chl'))) {
      return undefined;
    }
    const topic = url.pathname.match(/^\/t\/(?:([^/]+)\/)?([1-9][0-9]*)(?:\/.*)?$/);
    if (topic) {
      url.pathname = `/t/${topic[1] || 'topic'}/${topic[2]}`;
      url.search = '';
    } else if (/^\/(?:latest|top|categories)\/?$/i.test(url.pathname)) {
      url.pathname = url.pathname.replace(/\/$/, '');
      url.search = '';
    }
    url.hash = '';
    return url.toString();
  } catch {
    return undefined;
  }
}

module.exports = {
  MAX_ENTRIES,
  actionTitle,
  actionUrl,
  addHistoryEntry,
  createHistoryEntry,
  normalizePublicUrl,
  normalizeStoredHistory
};
